quixstreams/sources/community/file/origins/s3.py

```diff
@@ -0,0 +1,84 @@
+import logging
+from io import BytesIO
+from os import getenv
+from pathlib import Path
+from typing import Generator, Optional, Union
+
+from .base import ExternalOrigin
+
+try:
+    from boto3 import client as boto_client
+    from mypy_boto3_s3 import S3Client
+except ImportError as exc:
+    raise ImportError(
+        f"Package {exc.name} is missing: "
+        'run "pip install quixstreams[s3]" to use S3Origin'
+    ) from exc
+
+logger = logging.getLogger(__name__)
+
+__all__ = ("S3Origin",)
+
+
+class S3Origin(ExternalOrigin):
+    def __init__(
+        self,
+        bucket: str,
+        aws_region: Optional[str] = getenv("AWS_REGION"),
+        aws_access_key_id: Optional[str] = getenv("AWS_ACCESS_KEY_ID"),
+        aws_secret_access_key: Optional[str] = getenv("AWS_SECRET_ACCESS_KEY"),
+        aws_endpoint_url: Optional[str] = getenv("AWS_ENDPOINT_URL_S3"),
+    ):
+        """
+        Configure IcebergSink to work with AWS Glue.
+
+        :param bucket: The S3 bucket name only (ex: 'your-bucket').
+        :param aws_region: The AWS region.
+            NOTE: can alternatively set the AWS_REGION environment variable
+        :param aws_access_key_id: the AWS access key ID.
+            NOTE: can alternatively set the AWS_ACCESS_KEY_ID environment variable
+        :param aws_secret_access_key: the AWS secret access key.
+            NOTE: can alternatively set the AWS_SECRET_ACCESS_KEY environment variable
+        :param aws_endpoint_url: the endpoint URL to use; only required for connecting
+        to a locally hosted S3.
+            NOTE: can alternatively set the AWS_ENDPOINT_URL_S3 environment variable
+        """
+        self.root_location = bucket
+        self._credentials = {
+            "region_name": aws_region,
+            "aws_access_key_id": aws_access_key_id,
+            "aws_secret_access_key": aws_secret_access_key,
+            "endpoint_url": aws_endpoint_url,
+        }
+        # S3 client runs into pickling errors with multiprocessing. We can't set it
+        # until multiprocessing starts it.
+        # We can work around it by setting it during file collection
+        self._client: Optional[S3Client] = None
+
+    def _get_client(self) -> S3Client:
+        return boto_client("s3", **self._credentials)
```

Comment:
I think it should be `self._get_client` too.

BTW, maybe you could replace the `_get_client` with a `_client` property and a good explanation as to why it reinstantiates a client every time. Then `self._client: Optional[S3Client] = None` may be removed from init.

```suggestion
    @property
    def _client(self) -> S3Client:
        return boto_client("s3", **self._credentials)
```

Reply:
Well, it doesn't need to instantiate one every time...it basically must be instantiated _after_ it's kicked off as a `multiprocess` Process via `Source.run()` (which is where the pickling error is occurring).

The only reason this is a problem at all is I need to count the partition folders to see what the expected partition count is, which occurs before actually performing the run operation, so that's the thing I have to work around (by not storing that instance of the client when I do that call). 

Ideally I'd like to not have to instantiate a new client every time I call it...currently I'm setting the client on the first call I know that is made and then I can just re-use it as originally desired.